crates/replication2/src/coordinator/mod.rs

```diff
@@ -0,0 +1,286 @@
+        let operation = &operation;
+
+        let key = hash(operation.key());
+        let is_write = operation.is_write();
+
+        let cluster_view = &self.cluster.view();
+
+        let mut primary_quorum = Quorum::new(cluster_view.primary_replica_set(key));
+        let primary_replicas = primary_quorum.replica_set;
+        let primary_replica_requests = primary_quorum
+            .replica_set
+            .map(|operator| execute(operator, operation).map(move |resp| (operator, resp)))
+            .map(stream::once)
+            .merge();
+
+        let is_primary_replica =
+            |operator: &NodeOperator<_>| primary_replicas.iter().any(|op| op.id == operator.id);
+
+        // If there's an ongoing data migration and this is a write, then we need to
+        // replicate to an additional set of replicas.
+        let mut secondary_quorum = is_write
+            .then(|| cluster_view.secondary_replica_set(key).map(Quorum::new))
+            .flatten();
+
+        // Make sure that if an operator is in both primary and secondary
+        // keyspaces we only replicate once.
+        let send_secondary_replica_request = |operator| async move {
+            if is_primary_replica(operator) {
+                return None;
+            }
+
+            Some((operator, execute(operator, operation).await))
+        };
+
+        let secondary_replica_requests = (&secondary_quorum).pipe(|opt| {
+            let Some(quorum) = opt else {
+                return Either::Left(stream::empty());
+            };
+
+            quorum
+                .replica_set
+                .map(send_secondary_replica_request)
+                .map(stream::once)
+                .merge()
+                .filter_map(future::ready)
+                .pipe(Either::Right)
+        });
+
+        let mut response_futures =
+            pin!((primary_replica_requests, secondary_replica_requests).merge());
+
+        let mut responses: ResponseBuffer = std::array::from_fn(|_| None);
+
+        let quorum_response_idx = loop {
+            let Some((operator_idx, response)) = response_futures.next().await else {
+                break None;
+            };
+
+            let response_idx = receive_response(&mut responses, response);
+
+            let primary_quorum_response =
+                primary_quorum.response_received(operator_idx, response_idx);
+
+            let secondary_quorum_response = secondary_quorum
+                .as_mut()
+                .and_then(|replica_set| replica_set.response_received(operator_idx, response_idx));
+
+            match (primary_quorum_response, secondary_quorum_response) {
+                (Some(idx), _) if secondary_quorum.is_none() => break Some(idx),
+
+                // If both replica sets reached quorum, but the responses are different, then we
+                // return `None` indicating that quorum hasn't been reached.
+                (Some(a), Some(b)) => break Some(a).filter(|_| a == b),
```

Comment:
Why not `if a == b { Some(a) } else { None }`?

Reply:
There was probably a more complicated chain at some point.